### PR TITLE
feat: engine type implementation (will replace node api preview feature)

### PR DESCRIPTION
**Pull request by williamluke4**
Both the CLI and Client are still by default using the `binary` but this allows for the change over to be much easier.  
#### ENV OVERRIDE
```
PRISMA_CLI_QUERY_ENGINE_TYPE='library' | 'binary'
```

### CODE REVIEW

File: packages/engines/src/index.ts [hunk 1]
```diff
@@ -6,18 +6,36 @@ const debug = Debug('prisma:engines')
 export function getEnginesPath() {
   return path.join(__dirname, '../')
 }
-
+export const DEFAULT_CLI_QUERY_ENGINE_BINARY_TYPE = BinaryType.queryEngine
```
Comment (Jolg42) [inline packages/engines/src/index.ts:9]: I see it's exported, do we plan to use somewhere?

File: packages/fetch-engine/src/download.ts [hunk 2]
```diff
@@ -275,10 +275,12 @@ async function binaryNeedsToBeDownloaded(
   version: string,
   failSilent?: boolean,
 ): Promise<boolean> {
-  const binaryPath = job.envVarPath ?? job.targetFilePath
-
+  // If there is an ENV Override and the file exists then it does not need to be downloaded
+  if (job.envVarPath && fs.existsSync(job.envVarPath)) {
+    return false
+  }
```
Comment (Jolg42) [inline packages/fetch-engine/src/download.ts:281]: Looks like an unrelated optimization

**Review by Jolg42 — APPROVED** [review]
Nice! 💯

**Review by janpio — APPROVED** [review]

**Review by millsp — APPROVED** [review]

**Review by janpio — APPROVED** [review]
👍 

Note: Change dependencies before merging.
